File: Python/libraries/recognizers-date-time/recognizers_date_time/date_time/chinese/duration_parser.py
from typing import Optional
from datetime import datetime

from recognizers_number_with_unit import NumberWithUnitParser

from ..constants import TimeTypeConstants
from ..utilities import DateTimeResolutionResult
from ..extractors import ExtractResult
from ..parsers import DateTimeParseResult
from ..base_duration import BaseDurationParser
from .duration_parser_config import ChineseDurationParserConfiguration, ChineseDurationNumberWithUnitParserConfiguration

class ChineseDurationParser(BaseDurationParser):
    def __init__(self):
        super().__init__(ChineseDurationParserConfiguration())
        self._internal_parser = NumberWithUnitParser(ChineseDurationNumberWithUnitParserConfiguration())

    def parse(self, source: ExtractResult, reference: datetime = None) -> Optional[DateTimeParseResult]:
        if reference is None:
            reference = datetime.now()

        result = DateTimeParseResult()

        if source.type is self.parser_type_name:
            inner_result = DateTimeResolutionResult()

            has_half_suffix = source.text.endswith('半')

            if has_half_suffix:
                source.length = source.length - 1
                source.text = source.text[:source.length]

            parse_result = self._internal_parser.parse(source)
            unit_result = parse_result.value

            if not unit_result:
                return result

            unit_str = unit_result.unit
            number_str = unit_result.number

            if has_half_suffix:
                number_str = str((float(number_str) + 0.5))

            unit_type = 'T' if self.is_less_than_day(unit_str) else ''
            time_value = int(float(number_str) * self.config.unit_value_map.get(unit_str, 1))

            inner_result.timex = f'P{unit_type}{number_str}{unit_str[0]}'
            inner_result.future_value = time_value
            inner_result.past_value = time_value
            inner_result.future_resolution[TimeTypeConstants.DURATION] = str(inner_result.future_value)
            inner_result.past_resolution[TimeTypeConstants.DURATION] = str(inner_result.past_value)

            result = DateTimeParseResult(source)
            result.value = inner_result
            result.timex_str = inner_result.timex if inner_result is not None else ''
            result.resolution_str = ''

        return result

File: .NET/Microsoft.Recognizers.Text.DateTime/Japanese/Parsers/DurationParser.cs
using System.Collections.Generic;
using System.Globalization;
using DateObject = System.DateTime;

using Microsoft.Recognizers.Definitions.Japanese;
using Microsoft.Recognizers.Text.NumberWithUnit.Japanese;
using Microsoft.Recognizers.Text.NumberWithUnit;

using static Microsoft.Recognizers.Text.DateTime.Japanese.DurationExtractor;

namespace Microsoft.Recognizers.Text.DateTime.Japanese
{
    public class DurationParser : IDateTimeParser
    {
        public static readonly string ParserName = Constants.SYS_DATETIME_DURATION; //"Duration";

        private static readonly IParser InternalParser = new NumberWithUnitParser(new DurationParserConfiguration());

        internal class DurationParserConfiguration : JapaneseNumberWithUnitParserConfiguration
        {
            public DurationParserConfiguration() : base(new CultureInfo(Culture.Japanese))
            {
                this.BindDictionary(DurationExtractorConfiguration.DurationSuffixList);
            }
        }

        public static readonly Dictionary<string, int> UnitValueMap = DateTimeDefinitions.DurationUnitValueMap;

        private readonly IFullDateTimeParserConfiguration config;

        public DurationParser(IFullDateTimeParserConfiguration configuration)
        {
            config = configuration;
        }

        public ParseResult Parse(ExtractResult extResult)
        {
            return this.Parse(extResult, DateObject.Now);
        }

        public DateTimeParseResult Parse(ExtractResult er, DateObject refDate)
        {
            var referenceTime = refDate;

            // handle cases like "三年半"
            var hasHalfSuffix = false;
            if (er.Text.EndsWith("半"))
            {
                er.Length -= 1;
                er.Text = er.Text.Substring(0, er.Text.Length - 1);
                hasHalfSuffix = true;
            }

            var parseResult = InternalParser.Parse(er);
            var unitResult = parseResult.Value as UnitValue;

            if (unitResult == null)
            {
                return null;
            }

            var dtParseResult = new DateTimeResolutionResult();
            var unitStr = unitResult.Unit;
            var numStr = unitResult.Number;

            if (hasHalfSuffix)
            {
                numStr = (double.Parse(numStr) + 0.5).ToString(CultureInfo.InvariantCulture);
            }

            dtParseResult.Timex = "P" + (BaseDurationParser.IsLessThanDay(unitStr) ? "T" : "") + numStr + unitStr[0];
            dtParseResult.FutureValue = dtParseResult.PastValue = double.Parse(numStr) * UnitValueMap[unitStr];
            dtParseResult.Success = true;

            if (dtParseResult.Success)
            {
                dtParseResult.FutureResolution = new Dictionary<string, string>
                {
                    {TimeTypeConstants.DURATION, dtParseResult.FutureValue.ToString()}
                };

                dtParseResult.PastResolution = new Dictionary<string, string>
                {
                    {TimeTypeConstants.DURATION, dtParseResult.PastValue.ToString()}
                };
            }

            var ret = new DateTimeParseResult
            {
                Text = er.Text,
                Start = er.Start,
                Length = er.Length,
                Type = er.Type,
                Data = er.Data,
                Value = dtParseResult,
                TimexStr = dtParseResult.Timex,
                ResolutionStr = ""
            };

            return ret;
        }

        public List<DateTimeParseResult> FilterResults(string query, List<DateTimeParseResult> candidateResults)
        {
            return candidateResults;
        }

    }
}